php/libraries/NDB_Caller.class.inc

```diff
@@ -480,26 +493,29 @@ class NDB_Caller
             );
         }
 
-        // save instrument form data
-        $success = $instrument->save();
+        $status = new NDB_BVL_InstrumentStatus;
+        $status->select($commentID);
 
-        if ($redirectToOnSuccess !== null && $success !== false) {
-            return "<script language='javascript'>"
-                .   "window.location='$redirectToOnSuccess'"
-                . "</script>";
+        $has_read_permission = $user->hasPermission("access_all_profiles")
+            || $user->hasCenter($timepoint->getCenterID());
+        if (!$has_read_permission) {
+            throw new Exception("You cannot read this instrument");
         }
-        // create an instrument status object
-        $status  = new NDB_BVL_InstrumentStatus;
-        $success = $status->select($commentID);
-
+        $has_write_permission = $has_read_permission
+            && $user->hasCenterPermission("data_entry", $timepoint->getCenterID());
         // freeze the form to prevent data entry
-        if ($status->getDataEntryStatus() != 'In Progress'
-            || !$user->hasPermission('data_entry')
-            || !in_array(
-                $timepoint->getData("CenterID"),
-                $user->getData("CenterIDs")
-            )
+        if ($status->getDataEntryStatus() == 'In Progress'
+            && $has_write_permission
         ) {
+            if (strtoupper($_SERVER["REQUEST_METHOD"]) === "POST") {
+                $success = $instrument->save();
+                if ($redirectToOnSuccess != null && $success !== false) {
+                    return "script language='javascript'>"
```

Comment:
This PR should be rebased since #2778 has just been merged. Also please change this redirect to use `header` instead. Thanks!